File: codingTest/3xn타일링.py

def solution(n):
    answer = [0,3,11]
    index = int(n/2)
    if n % 2 != 0: return 0
    if index < 3: return answer[index]

    for i in range(3, index+1):
    	# answer[i:j] -> answer에서 index가 i인 원소부터 j-1인 원소까지의 sub-array
        answer.append((3*answer[i-1]+sum(answer[1:i-1])*2+2)%1000000007)

    return answer[index]



print(solution(12))

File: codingTest/3xn타일링.js
function solution(n) {
    var answer = 0;
    const dp = Array.from({length: n + 1}, () => 0);
    dp[2] = 3;
    dp[3] = 0;
    dp[4] = 11;
    
    for (let i = 5; i < n + 1; i++) {
        if (i % 2 === 1) {
            dp[i] = 0
        }
        else if (i % 2 === 0) {
            dp[i] += dp[i - 2] * 3
            for (let j = 2; j < i - 2; j += 2) {
                dp[i] += dp[j] * 2
            }
            dp[i] += 2
            dp[i] = dp[i] % 1000000007
        }
    }
    
    // console.log(dp);
    return dp[n] % 1000000007;
}
